File: frontend/lib/models/message.dart
import 'package:json_annotation/json_annotation.dart';

part 'message.g.dart';

@JsonSerializable()
class Message {
  final String id;
  final String content;
  final bool encrypted;
  final String channelId;
  final String authorId;
  final String authorUsername;
  final String? authorAvatar;
  final String? replyToId;
  final DateTime? editedAt;
  final DateTime? createdAt;

  Message({
    required this.id,
    required this.content,
    this.encrypted = false,
    required this.channelId,
    required this.authorId,
    required this.authorUsername,
    this.authorAvatar,
    this.replyToId,
    this.editedAt,
    this.createdAt,
  });

  factory Message.fromJson(Map<String, dynamic> json) => _$MessageFromJson(json);
  Map<String, dynamic> toJson() => _$MessageToJson(this);
}

@JsonSerializable()
class DMMessage {
  final String id;
  final String content;
  final bool encrypted;
  final String conversationId;
  final String authorId;
  final String authorUsername;
  final String? authorAvatar;
  final bool read;
  final DateTime? createdAt;

  DMMessage({
    required this.id,
    required this.content,
    this.encrypted = false,
    required this.conversationId,
    required this.authorId,
    required this.authorUsername,
    this.authorAvatar,
    this.read = false,
    this.createdAt,
  });

  factory DMMessage.fromJson(Map<String, dynamic> json) => _$DMMessageFromJson(json);
  Map<String, dynamic> toJson() => _$DMMessageToJson(this);
}

@JsonSerializable()
class DMConversation {
  final String id;
  final String user1Id;
  final String user2Id;
  final String otherUserId;
  final String otherUsername;
  final String? otherAvatar;
  final bool otherIsOnline;
  final DMMessage? lastMessage;
  final int unreadCount;
  final DateTime? createdAt;

  DMConversation({
    required this.id,
    required this.user1Id,
    required this.user2Id,
    required this.otherUserId,
    required this.otherUsername,
    this.otherAvatar,
    this.otherIsOnline = false,
    this.lastMessage,
    this.unreadCount = 0,
    this.createdAt,
  });

  factory DMConversation.fromJson(Map<String, dynamic> json) => _$DMConversationFromJson(json);
  Map<String, dynamic> toJson() => _$DMConversationToJson(this);
}

File: backend/app/schemas/message.py
from pydantic import BaseModel, Field
from typing import Optional, List
import uuid
from datetime import datetime


class MessageCreate(BaseModel):
    content: str = Field(..., min_length=1, max_length=4000)
    encrypted: bool = False
    reply_to_id: Optional[uuid.UUID] = None


class MessageUpdate(BaseModel):
    content: str = Field(..., min_length=1, max_length=4000)


class MessageResponse(BaseModel):
    id: uuid.UUID
    content: str
    encrypted: bool
    channel_id: uuid.UUID
    author_id: uuid.UUID
    author_username: str
    author_avatar: Optional[str] = None
    reply_to_id: Optional[uuid.UUID] = None
    edited_at: Optional[datetime] = None
    created_at: datetime

    class Config:
        from_attributes = True


class DMMessageCreate(BaseModel):
    content: str = Field(..., min_length=1, max_length=4000)
    encrypted: bool = False


class DMMessageResponse(BaseModel):
    id: uuid.UUID
    content: str
    encrypted: bool
    conversation_id: uuid.UUID
    author_id: uuid.UUID
    author_username: str
    author_avatar: Optional[str] = None
    read: bool
    created_at: datetime

    class Config:
        from_attributes = True


class DMConversationResponse(BaseModel):
    id: uuid.UUID
    user1_id: uuid.UUID
    user2_id: uuid.UUID
    other_user_id: uuid.UUID
    other_username: str
    other_avatar: Optional[str] = None
    other_is_online: bool
    last_message: Optional[DMMessageResponse] = None
    unread_count: int = 0
    created_at: datetime

    class Config:
        from_attributes = True
